ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -563,17 +567,109 @@ instance Serialise CC.Common.KeyHash where
   encode = toCBOR
   decode = fromCBOR
 
-encodeByronHeader
-  :: Crypto.ProtocolMagicId
-  -> CC.Slot.EpochSlots
-  -> Header (ByronBlockOrEBB cfg)
-  -> Encoding
-encodeByronHeader pm epochSlots (ByronHeaderOrEBB h) = toCBORAHeaderOrBoundary pm epochSlots h
+-- Codec for blocks and headers using CBOR-in-CBOR. This style is needed
+-- because the cardano-ledger decoders drop information that must be
+-- retained if we are to communicate with Byron peers.
+-- Re-annotating in the cardano-ledger style, i.e. re-encoding then using those
+-- bytes to annotate, does not work, because information is lost in the
+-- initial decoding.
+-- This codec must be used for network exchange _and_ for the database.
+
+-- | Get the encoded bytes of a block. A legacy Byron node (cardano-sl) would
+-- successfully decode a block from these.
+blockBytes :: ByronBlockOrEBB cfg -> Lazy.ByteString
+blockBytes blk = Lazy.pack [listLengthByte, discriminatorByte] `Lazy.append` Lazy.fromStrict mainBytes
+  where
+  -- Here's how it would look if we could use CBOR encoding.
+  --
+  --  enc = CBOR.encodeListLen 2 <> blockBytes
+  --  blockEnc = case unByronBlockOrEBB blk of
+  --    CC.Block.ABOBBoundary b -> toCBOR (0 :: Word) <> toCBOR
+  --    CC.Block.ABOBBlock b    -> toCBOR (1 :: Word) <> toCBORBlock epochSlots b
+  --
+  -- But since we don't carry an annotation for the _entire_ block, and we
+  -- can't put a bytestring representing a _part_ of the encoding into an
+  -- encoding, we have to work directly with bytes. Here's how it goes:
+  --
+  --   0b100_00010 for list length (major type 4) value 2
+  --   0b000_0000x where x is 1 for EBB, 0 for main block. Unsigned integer (major type 0)
+  --
+  -- No binary literals so we'll write them hex.
+  listLengthByte = 0x82
+  discriminatorByte = case unByronBlockOrEBB blk of
+    CC.Block.ABOBBoundary _ -> 0x00
+    CC.Block.ABOBBlock    _ -> 0x01
+  mainBytes = case unByronBlockOrEBB blk of
+    CC.Block.ABOBBlock    b -> CC.Block.blockAnnotation b
+    CC.Block.ABOBBoundary b -> CC.Block.boundaryAnnotation b
+
+-- | Encode a block using CBOR-in-CBOR tag 24.
+encodeByronBlock :: ByronBlockOrEBB cfg -> Encoding
+encodeByronBlock = encodeUnknownCborDataItem . blockBytes
```

Comment:
We should not use `encodeUnknownCborDataItem` even though it uses the right CBOR encoding. It carries misleading semantic baggage. Its name comes from the old codebase where we had to support unknown values as an extension mechanism. These two functions ought to be banished to the byron legacy support and not be used for anything else.

If we want to use special utils for this then we need new ones with appropriate names and documentation.